### PR TITLE
N°6808 - Rank management (order) in iTop actions

### PR DESCRIPTION
**Pull request by accognet**
Internal

### CODE REVIEW

**Comment by Molkobain** [comment]
Instead of sorting element in PHP, did you check with @rquetiez if we can specify the sort order on the `lnkTriggerAction` class (`/itop_design/classes/class/properties/order`) so elements are sorted when retrieved from the DB?

**Comment by accognet** [comment]
I have tested by adding the "order" attribute to the lnk and the sorting is not used.

**Comment by Molkobain** [comment]
Accepted during functional review, but we need to check whether the different actions are executed in sequence (wait for the previous one to be done before the next one is executed) or if they are in //. It should be in sequence but there was a confusion during the meeting.

Implementation will be discussed during next technical review.

**Comment by Molkobain** [comment]
Accepted during technical review once sort is done naturally to avoid falling in the usual pitfall "1, 10, 2" order.

> Accepted during functional review, but we need to check whether the different actions are executed in sequence (wait for the previous one to be done before the next one is executed) or if they are in //. It should be in sequence but there was a confusion during the meeting.

Also, actions are executed sequentially.

File: core/trigger.class.inc.php [hunk 1]
```diff
@@ -113,17 +113,22 @@ public function DoActivate($aContextArgs)
...
-			if ($oAction->IsActive())
-			{
-                $oKPI = new ExecutionKPI();
-				$oAction->DoExecute($this, $aContextArgs);
-                $oKPI->ComputeStatsForExtension($oAction, 'DoExecute');
+		$aActionListOrdered = [];
+		while ($oLink = $oLinkedActions->Fetch()) {
+			$aActionListOrdered[$oLink->Get('order')][] = $oLink;
+		}
+		ksort($aActionListOrdered);
```
Comment (Molkobain) [inline core/trigger.class.inc.php:120]: Use `natsort()` to avoid actions being executed like "1, 10, 2"

File: core/trigger.class.inc.php [hunk 2]
```diff
@@ -113,17 +113,22 @@ public function DoActivate($aContextArgs)
...
-			/** @var \Action $oAction */
-			$oAction = MetaModel::GetObject('Action', $iActionId);
-			if ($oAction->IsActive())
-			{
-                $oKPI = new ExecutionKPI();
-				$oAction->DoExecute($this, $aContextArgs);
-                $oKPI->ComputeStatsForExtension($oAction, 'DoExecute');
+		$aActionListOrdered = [];
+		while ($oLink = $oLinkedActions->Fetch()) {
+			$aActionListOrdered[$oLink->Get('order')][] = $oLink;
```
Comment (Molkobain) [inline core/trigger.class.inc.php]:
```suggestion
			$aActionListOrdered[(int) $oLink->Get('order')][] = $oLink;
```

To ensure that order is always returned as an int, so ksort work even if Get() starts returning strings.

**Review by Molkobain — APPROVED** [review]

File: core/trigger.class.inc.php [hunk 3]
```diff
@@ -113,17 +113,22 @@ public function DoActivate($aContextArgs)
...
-			/** @var \Action $oAction */
-			$oAction = MetaModel::GetObject('Action', $iActionId);
-			if ($oAction->IsActive())
-			{
-                $oKPI = new ExecutionKPI();
-				$oAction->DoExecute($this, $aContextArgs);
-                $oKPI->ComputeStatsForExtension($oAction, 'DoExecute');
+		$aActionListOrdered = [];
+		while ($oLink = $oLinkedActions->Fetch()) {
+			$aActionListOrdered[(int) $oLink->Get('order')][] = $oLink;
```
Comment (Hipska) [inline core/trigger.class.inc.php:121]: ~~Erm, I always thought `order` could be any numeric, so also float?~~

Nevermind, I was confusing with `<rank>`.

File: core/trigger.class.inc.php [hunk 4]
```diff
@@ -113,17 +113,22 @@ public function DoActivate($aContextArgs)
...
+		}
+		ksort($aActionListOrdered);
+		foreach ($aActionListOrdered as $aActionSubList) {
```
Comment (Molkobain) [inline core/trigger.class.inc.php]:
```suggestion

		// Order actions as expected
		$aActionListOrdered = [];
		while ($oLink = $oLinkedActions->Fetch()) {
			$aActionListOrdered[(int) $oLink->Get('order')][] = $oLink;
		}
		ksort($aActionListOrdered);
		
		// Execute actions
		foreach ($aActionListOrdered as $aActionSubList) {
```

**Comment by Molkobain** [comment]
All good! You can:
- Rebase branch on support/3.2
- Make sure target branch is support/3.2
- Squash and merge (mind to keep the PR number in the title)

**Pull request merged**